### DISCUSSION TITLE
Best practice for new Settings

### DISCUSSION THREAD

**Discussion by sillyfrog** [arendst/Tasmota#12500] [Q&A] [answered]
I have had a bit of a look, but unable to find the "best practice" for adding new settings. I'm hoping to add the ability to add an AP to a Tasmota instance to act as a WiFi Range Extender / Repeater (see #12421 for more info). 

The issue with this enhancement is it integrates with the WiFi portion, so it's not like a typical "driver" that's adding a new sensor, so hoping for some pointers as to the best way to get these options from an operator and use them in my code.

Thanks!

**Comment by barbudor** [comment, marked as answer]
Hi @sillyfrog 

### Short answer:
I would advise a 2 steps approach:
- 1st step, design your feature without persistent storage which means that settings must be redone at each reboot. That can be easily performed using a rule such as `on system#boot do backlog xxx ; xxx ; xxx endon`, or using `autoexec.bat` if you can have filesystem.
- 2nd step, depending on the traction and usage of your feature, Theo will see if fixed room can be allocated in the flash settings

### Longuer answer:

Persistent settings need to be located in the `TSettings` structure of [Settings.h](https://github.com/arendst/Tasmota/blob/development/tasmota/settings.h). To add a new setting, it is important to keep the structure and offset location of existing settings stable. So when loading a new code on a device with existing settings area, it remains compliant.

You need to identify some free bytes in the structure as identified by `uint8_t free_xxxx[nnn];` where `xxx` is the byte-offset (in hex) from the beginning of the structure and `nnn` the number of bytes. Then insert your settings there, preserving offset and alignments.
For example, let's say you want to add `uint16_t sillyfrog_data;`, you could change:
```
  uint8_t       switchmode[MAX_SWITCHES_SET];  // 4A9
  uint8_t       free_4c5[5];               // 4C5
  uint8_t       ex_interlock[4];           // 4CA MAX_INTERLOCKS = MAX_RELAYS / 2 (Legacy)
```
into
```
  uint8_t       switchmode[MAX_SWITCHES_SET];  // 4A9
  uint8_t       free_4c5[3];               // 4C5
  uint16_t     sillyfrog_data;           // 4C8 SillyFrog data
  uint8_t       ex_interlock[4];           // 4CA MAX_INTERLOCKS = MAX_RELAYS / 2 (Legacy)
```
Note the offset, as the new fields needs to be aligned modulo its size (uint16_t aligned on modulo 2, uint32_t aligned on modulo 4).
It is even more complex in a structure as if you don't take care, the compiler may re-align fields in the structure and loose bytes.

However, flash setting is rare and precious and usage is subject to approval by Theo. Especially if you require a large area such as storing an AP name. Don't forget that settings are assigned once for all and use the place even if the feature is not used or even not build into a given binary.
There is a special place for strings (`TSettings->text_pool[]`) where all strings are stored concatenated, `'\0'` separated, which would use only 1 byte if the string is not set, but the room is not so large and availability depends on usage of other features.

If your feature is meant to run only on devices with >1MB that can hold a filesystem, you can also look at [`xdrv_98_file_settings_demo.ino`](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_98_file_settings_demo.ino) which demonstrate how to store driver's settings into a file instead of the global flash-settings area.

**Reply by sillyfrog** [reply]
Thanks! Exactly what I need to know. I think I'll just leave it with `#define`s for the moment as taking up flash space even when not included is not ideal for something that won't be used so much.

**Reply by barbudor** [reply]
You can still implement configuration commands, even if they are not stored persistent.
That would make the feature usable without having to recompile to change a setting

**Reply by sillyfrog** [reply]
Do you have any pointers for how to do that? Even just an example in the code I could reference? 

Thanks again!

**Reply by barbudor** [reply]
Almost any module/driver implements commands so there's a lot of example.
However you can have a look at [xdrv_98_file_settings_demo.ino](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_98_file_settings_demo.ino)
It shows how to implement a command "DrvText" where "Drv" is a prefix within the function `CmndDrvText()` and the data are saved into a private variable `Drv98Settings`.
If you ignore the part about saving the settings into a file, you have the core of a setting-command.
